### PR TITLE
Re-enable some lints from app and app-channels

### PR DESCRIPTION
**Pull request by arnauorriols**
# Description of change

Some Clippy lints were disabled temporarily to favor speed of delivery of needed features. This PR refactors the several parts of the code to re-enable some of those lints.   

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

Specially relevant breaking changes:
- Messages are now indexed using the binary digest of the `Address` blake2b256 hash. Messages were previously hashed using the hexadecimal encoding of said hash. Warning: old messages will **not** be compatible with this new version.
- `Address` `Display` format now outputs the same as `Address::to_string()`. To get the hash of the address one must use `Address::to_blake2b()` or `Address::to_msg_index()`

## How the change has been tested
Run the following examples:

- streams/examples
- bindings/wasm examples
- bindings/c examples
- iota-streams-app-channels-example

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -3,7 +3,7 @@
 [package]
 name = "iota-streams"
 version = "1.0.0"
-authors = ["Vlad Semenov <vlad.semenov@iota.org>", "Dyrell Chapman <dyrell.chapman@iota.org>", "Brord van Wierst <brord@iota.org>"]
+authors = ["Vlad Semenov <vlad.semenov@iota.org>", "Dyrell Chapman <dyrell.chapman@iota.org>", "Brord van Wierst <brord@iota.org>", "Arnau Orriols <arnau.orriols@iota.org"]
```
Comment (kwek20) [inline Cargo.toml:6]: Welcome!

File: iota-streams-app/Cargo.toml [hunk 2]
```diff
@@ -41,7 +41,7 @@ async-trait = { version = "0.1", optional = true }
 atomic_refcell = { version = "0.1.6", optional = true }
 
 # Dependencies for "client" feature
-iota-client = { git = "https://github.com/iotaledger/iota.rs", branch = "dev", default-features = false, optional = true }
+iota-client = { git = "https://github.com/iotaledger/iota.rs", rev  = "283af58b6f9a20b3698e5128e3e651b2a4dce861", default-features = false, optional = true }
```
[the remark this answers is not in the document]
Comment (arnauorriols) [reply iota-streams-app/Cargo.toml:44]: wasm is broken in iota.rs, didn't want to spend time bisecting their git log. This commit is the one that was pinned in the `Cargo.lock`, therefore it was the one being used right until we removed the `Cargo.lock`, therefore certain to work with wasm.

File: bindings/c/src/api/mod.rs [hunk 3]
```diff
@@ -526,44 +521,44 @@ pub extern "C" fn drop_payloads(payloads: PacketPayloads) {
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn drop_str(string: *const c_char) {
+pub unsafe extern "C" fn drop_str(string: *mut c_char) {
```
Comment (semenov-vladyslav) [inline bindings/c/src/api/mod.rs:524]: Why `mut` here? The idea was that `drop_str` frees strings returned by bindings/c API which is of type `*const c_char`, ie. immutable strings.

File: bindings/c/src/api/mod.rs [hunk 4]
```diff
@@ -526,44 +521,44 @@ pub extern "C" fn drop_payloads(payloads: PacketPayloads) {
...
-pub unsafe extern "C" fn get_address_index_str(address: *mut Address) -> *mut c_char {
-    address.as_ref().map_or(null_mut(), |addr| {
-        get_hash(addr.appinst.as_ref(), addr.msgid.as_ref()).map_or(null_mut(), |index| {
-            CString::new(index).map_or(null_mut(), |index| index.into_raw())
-        })
+pub unsafe extern "C" fn get_address_index_str(address: *const Address) -> *const c_char {
+    address.as_ref().map_or(null(), |addr| {
+        let index = addr.to_msg_index();
+        let index_hex = format!("{:x}", index);
+        CString::new(index_hex).map_or(null(), |index| index.into_raw())
```
Comment (semenov-vladyslav) [inline bindings/c/src/api/mod.rs:561]: Maybe factor this pattern out into a separate function `string_into_raw` and replace other occurrences (if any)?

File: iota-streams-app-channels/src/api/key_store.rs [hunk 5]
```diff
@@ -64,8 +69,8 @@ impl<Info, F: PRP> KeyStore<Info, F> for KeyMap<Info> {
                 Identifier::PskId(_id) => self
                     .psks
                     .get_key_value(id)
-                    .filter(|(_, (x, _))| x.is_some())
-                    .map(|(e, (x, _))| (e, x.unwrap().to_vec())),
+                    .map(|(e, (x, _))| x.map(|xx| (e, xx.to_vec())))
```
Comment (semenov-vladyslav) [inline iota-streams-app-channels/src/api/key_store.rs:72]: Maybe `filter_map` would be more suitable?

**Review by semenov-vladyslav — APPROVED** [review]

File: Cargo.toml (hunk 1 above)
Comment (arnauorriols) [reply Cargo.toml:6]: :D

File: bindings/c/src/api/mod.rs (hunk 3 above)
Comment (arnauorriols) [reply bindings/c/src/api/mod.rs:524]: You are right, don't know what I was thinking. Then I forgot to change the header anyhow, so the example run fine. Actually, I assumed we used `cbindgen` for that. Any reason why we are manually writing the header instead of generating them with `cbindgen`?

File: iota-streams-app-channels/src/api/key_store.rs (hunk 5 above)
Comment (arnauorriols) [reply iota-streams-app-channels/src/api/key_store.rs:72]: I thought about it, yeah. However, it does not really suit the scenario. `Option<T>` does not implement `filter_map()` as an specialized method (like it does with `map()`). Therefore, you need to convert the `Option` to an iterator first, polluting the pipeline with `into_iter()` and `collect()`. However, `Option<T>` only implements [`FromIterator<Item=Option<A>> where T: IntoIterator<A>`](https://doc.rust-lang.org/std/option/enum.Option.html#impl-FromIterator%3COption%3CA%3E%3E), which means you cannot build an `Option` out of an iterator unless you store all the items from the iterator in a collection inside the Option, which is not the case. We should probably introduce a `fold()` instead...

Long story short, `flatten()` is the nicest pattern I could come up with (and it's not uncommon).

File: bindings/c/src/api/mod.rs (hunk 4 above)
Comment (arnauorriols) [reply bindings/c/src/api/mod.rs]: Nice suggestion. Done!

File: iota-streams-app-channels/src/lib.rs [hunk 6]
```diff
@@ -29,3 +29,12 @@ pub mod message;
 
 /// Author and Subscriber API.
 pub mod api;
+
+/// Get a `Value` given a `Key`
+///
+/// This trait can be implemented to any kind of collection to get an item out of it.
+/// It's meant to be versatile, so it can be implemented for T or &T, and both `Key` and
+/// `Value` can be owned or references as well.
+pub trait Lookup<Key, Value> {
```
[the remark this answers is not in the document]
Comment (arnauorriols) [reply iota-streams-app-channels/src/lib.rs:38]: With something so "silly" as this it didn't even occur to me to search for an exernal implementation... Why would you consider it? 

The left-pad incident always comes to mind when thinking about these kinds of decisions... :joy: Anyway, in Rust this also manifests as a bad idea because of the non-trivial build time introduced unnecessarily, for example, or unnecessary system-dependant compatibility issues from the new dependency and its own dependencies... just to name a few.

File: bindings/c/CMakeLists.txt [hunk 7]
```diff
@@ -4,8 +4,8 @@ project(iota_streams_c LANGUAGES C)
 include(FetchContent)
 
 option(NO_STD "Enable no_std build, without iota_client" OFF)
-option(SYNC_CLIENT "Enable sync transport via iota_client" OFF)
-option(STATIC "Build static library" ON)
+option(SYNC_CLIENT "Enable sync transport via iota_client" ON)
+option(STATIC "Build static library" OFF)
```
Comment (DyrellC) [inline bindings/c/CMakeLists.txt:8]: :pray:

File: bindings/c/src/api/mod.rs [hunk 8]
```diff
@@ -61,6 +58,24 @@ pub(crate) fn safe_drop_mut_ptr<T>(p: *mut T) {
     }
 }
 
+/// Convert an String-like collection of bytes into a raw pointer to the first byte
+///
+/// The pointer might be [`null`] if the String contains a null byte (which is invalid)
+///
+/// [`null`]: https://doc.rust-lang.org/std/ptr/fn.null.html
+fn _string_into_raw(string: impl Into<Vec<u8>>) -> *const c_char {
```
Comment (DyrellC) [inline bindings/c/src/api/mod.rs:66]: A good pair of utility functions :+1:

File: bindings/c/src/api/mod.rs [hunk 9]
```diff
@@ -533,37 +545,37 @@ pub unsafe extern "C" fn drop_str(string: *const c_char) {
...
+    address.as_ref().map_or(null(), |addr| get_msgid_str(&addr.msgid))
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn get_address_index_str(address: *mut Address) -> *mut c_char {
-    address.as_ref().map_or(null_mut(), |addr| {
-        get_hash(addr.appinst.as_ref(), addr.msgid.as_ref()).map_or(null_mut(), |index| {
-            CString::new(index).map_or(null_mut(), |index| index.into_raw())
-        })
+pub unsafe extern "C" fn get_address_index_str(address: *const Address) -> *const c_char {
```
Comment (DyrellC) [inline bindings/c/src/api/mod.rs:574]: Yeah this never needed to have mutable pointers, thanks for correcting that

File: bindings/c/src/api/auth.rs [hunk 10]
```diff
@@ -171,8 +171,10 @@ pub unsafe extern "C" fn auth_send_keyload(
             link_to.as_ref().map_or(Err::NullArgument, |link_to| {
                 psk_ids.as_ref().map_or(Err::NullArgument, |psk_ids| {
                     ke_pks.as_ref().map_or(Err::NullArgument, |ke_pks| {
-                        let identifiers: Vec<Identifier> = ke_pks.into_iter().map(|pk| (*pk).into()).collect();
-                        user.send_keyload(link_to, psk_ids, &identifiers.iter().collect())
+                        let pks = ke_pks.into_iter().copied().map(Into::<Identifier>::into);
```
Comment (DyrellC) [inline bindings/c/src/api/auth.rs:174]: Keep an eye out here, Vlad's PR changes this behaviour to have it all input as `Identifiers` instead of as separate collections of `Pk`'s and `PskId`'s. Depending on which order we pull these in we'll just have to make sure to fix it there appropriately. Ideally we should be putting through the other one first.

File: bindings/c/CMakeLists.txt (hunk 7 above)
Comment (DyrellC) [inline bindings/c/CMakeLists.txt:8]: :pray:

File: bindings/c/src/api/auth.rs (hunk 10 above)
Comment (DyrellC) [inline bindings/c/src/api/auth.rs:174]: This logic is going to change with Vlad's PR. We'll be using `Identifier's` directly instead of pulling in `Pk's` and `PskId's` respectively, they'll all have to be converted first and be injected as one unified collection.

File: bindings/c/src/api/mod.rs (hunk 8 above)
Comment (DyrellC) [inline bindings/c/src/api/mod.rs:66]: Fan of these utility functions :+1:

File: bindings/wasm/src/author/authorw.rs [hunk 11]
```diff
@@ -162,15 +162,16 @@ impl Author {
 
     #[wasm_bindgen(catch)]
     pub async fn send_keyload(self, link: Address, psk_ids: PskIdsW, sig_pks: PublicKeysW) -> Result<UserResponse> {
-        let pks: Vec<Identifier> = sig_pks.pks.iter().map(|pk| (*pk).into()).collect();
+        let pks = sig_pks.pks.into_iter().map(Into::<Identifier>::into);
```
Comment (DyrellC) [inline bindings/wasm/src/author/authorw.rs:165]: Same thing as before here, we'll be pulling in a collection of `Identifier's` directly. This will also change the `sig_pks` and `psk_ids` input types obviously, so we'll need to keep an eye on that and change them appropriately depending on the order of everything being accepted.

File: iota-streams-app-channels/src/api/key_store.rs [hunk 12]
```diff
@@ -15,7 +15,9 @@ use iota_streams_core::{
 use iota_streams_core_edsig::key_exchange::x25519;
 
 pub trait KeyStore<Info, F: PRP>: Default {
-    fn filter(&self, pks: &[&Identifier]) -> Vec<(&Identifier, Vec<u8>)>;
+    fn filter<'a, I>(&self, pks: I) -> Vec<(&Identifier, Vec<u8>)>
```
Comment (DyrellC) [inline iota-streams-app-channels/src/api/key_store.rs:18]: We should rename this from `pks` to `ids` or something, considering it can be either a pk or pskid (and in the future DID). It'll also match the actual implementation below that way too :sweat_smile:

File: iota-streams-app-channels/src/api/key_store.rs (hunk 5 above)
Comment (DyrellC) [reply iota-streams-app-channels/src/api/key_store.rs:72]: Yeah I originally tried to use a `filter_map` and ran into that exact scenario. Didn't consider a `fold()` though, that seems like a reasonable alternative.

File: iota-streams-app-channels/src/api/tangle/author.rs [hunk 13]
```diff
@@ -171,15 +171,12 @@ impl<Trans: Transport + Clone> Author<Trans> {
...
-    ///  * `ke_pks`  - Vector of Public Keys to be included in message
-    pub fn send_keyload(
-        &mut self,
-        link_to: &Address,
-        psk_ids: &PskIds,
-        ke_pks: &Vec<&Identifier>,
-    ) -> Result<(Address, Option<Address>)> {
-        self.user.send_keyload(link_to, psk_ids, ke_pks)
+    ///  * `keys`  - Iterable of [`Identifier`] to be included in message
+    pub fn send_keyload<'a, I>(&mut self, link_to: &Address, keys: I) -> Result<(Address, Option<Address>)>
```
Comment (DyrellC) [inline iota-streams-app-channels/src/api/tangle/author.rs:175]: Oh good you've already gone ahead and done this change here, we just need to update the bindings to reflect this same approach.

File: iota-streams-app-channels/src/api/tangle/user.rs [hunk 14]
```diff
@@ -554,10 +569,13 @@ impl<Trans: Transport + Clone> User<Trans> {
         ref_link: &MsgId,
         info: MsgInfo,
     ) -> Result<(Address, Option<Address>)> {
-        let seq = self.user.wrap_sequence(ref_link)?;
```
Comment (DyrellC) [inline iota-streams-app-channels/src/api/tangle/user.rs:557]: There was originally a reason why this was ordered this way but I can't recall the exact reason off the top of my head. If none of the tests failed along the way then whatever it was has likely been rectified already, but I just want to leave a note here noting that there was a purpose behind that at some point. In the meantime I'm content (even elated) with the reordering.

File: iota-streams-app-channels/src/api/user.rs [hunk 15]
```diff
@@ -58,66 +57,43 @@ use crate::{
...
-}
-
-impl<F: PRP, Link: HasLink + fmt::Debug> fmt::Debug for WrapStateSequence<F, Link>
+/// Sequence wrapping object
+///
+/// When using multibranch mode, this wrapping object contains the (wrapped) sequence message ([`WrappedMessage`]) to be
+/// sent and the [`Cursor`] of the user sending it.
+///
+/// When using single-branch mode, only the [`Cursor`] is needed, and no sequence message is sent.
+pub enum WrappedSequence<F, Link>
```
Comment (DyrellC) [inline iota-streams-app-channels/src/api/user.rs:73]: I think this is a good addition, it'll make single depth more succinct to implement as well.

File: iota-streams-app-channels/src/api/user.rs [hunk 16]
```diff
@@ -1311,3 +1243,34 @@ where
         Ok(user)
     }
 }
+
+impl<'a, F, Link, LG, LS, Keys> Lookup<&Identifier, psk::Psk> for &'a User<F, Link, LG, LS, Keys>
```
Comment (DyrellC) [inline iota-streams-app-channels/src/api/user.rs:1247]: :pray:

File: iota-streams-app-channels/src/message/keyload.rs [hunk 17]
```diff
@@ -107,23 +112,23 @@ where
...
-                                    .absorb(External(<&NBytes<psk::PskSize>>::from(<&[u8]>::from(&store_id))))?
-                                    .commit()?
-                                    .mask(&self.key),
-                                Identifier::EdPubKey(_pk) => ctx.x25519(
-                                    &x25519::PublicKey::from(<[u8; 32]>::try_from(store_id.as_ref())?),
-                                    &self.key,
-                                ),
-                            }
-                        })
+                ctx.absorb(repeated_keys)?.repeated(&self.keys, |ctx, (id, store_id)| {
```
Comment (DyrellC) [inline iota-streams-app-channels/src/message/keyload.rs:115]: Logic here is about to change https://github.com/iotaledger/streams/pull/125/files#diff-2a2e4f5ff0b3d34ef457c25d06f5fc2eaf1a4d0bc6feff728c56f5452e433af5R110 just so you know

File: iota-streams-app/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (DyrellC) [reply iota-streams-app/Cargo.toml:44]: It also works to use `branch = "wasm"` but we definitely are going to want to pressure that fix to be a part of dev sooner than later. And then pick a revision we feel comfortable with going forward.

File: iota-streams-app/src/transport/tangle/mod.rs [hunk 18]
```diff
@@ -138,30 +141,95 @@ impl<F> TangleMessage<F> {
 }
 
 /// Tangle representation of a Message Link
-#[derive(Clone)]
+///
```
[the remark this answers is not in the document]
Comment (DyrellC) [reply iota-streams-app/src/transport/tangle/mod.rs:145]: Seconded

File: iota-streams-app/src/transport/tangle/mod.rs [hunk 19]
```diff
@@ -138,30 +141,95 @@ impl<F> TangleMessage<F> {
...
-
-        let msgid =
-            MsgId::from_str(msgid_str).map_err(|e| wrapped_err!(BadHexFormat(appinst_str.into()), WrappedError(e)))?;
+    pub fn new(appinst: AppInst, msgid: MsgId) -> Self {
+        Self { appinst, msgid }
+    }
 
-        Ok(TangleAddress { appinst, msgid })
+    /// Hash the content of the TangleAddress using `Blake2b256`
+    pub fn to_blake2b(&self) -> NBytes<U32> {
```
[the remark this answers is not in the document]
Comment (DyrellC) [reply iota-streams-app/src/transport/tangle/mod.rs:223]: Since we have the `to_msg_index` function now I would agree

File: iota-streams-app/src/transport/tangle/mod.rs [hunk 20]
```diff
@@ -138,30 +141,95 @@ impl<F> TangleMessage<F> {
...
-    pub fn to_string(&self) -> String {
-        let mut address = String::new();
-        address.push_str(&self.appinst.to_string());
-        address.push(':');
-        address.push_str(&self.msgid.to_string());
-        address
+    /// Generate the hash used to index the [`TangleMessage`] published in this address
+    ///
+    /// Currently this hash is computed with [`to_blake2b`].
+    pub fn to_msg_index(&self) -> NBytes<U32> {
```
[the remark this answers is not in the document]
Comment (DyrellC) [reply iota-streams-app/src/transport/tangle/mod.rs:232]: Seems easy enough to just add in now

**Review by DyrellC — COMMENTED** [review]
Looking really good, couple of smaller comments, adding on to a couple statements Brord made. Depending on how Vlad's PR is looking we should be able to get this pulled in pretty soon here.

File: iota-streams-app/src/transport/tangle/mod.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (arnauorriols) [reply iota-streams-app/src/transport/tangle/mod.rs:232]: Working on it!

File: iota-streams-app/src/transport/tangle/mod.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (arnauorriols) [reply iota-streams-app/src/transport/tangle/mod.rs:223]: Yeah you are right. I first exposed this method, and later I came up with the more business-oriented `to_msg_index()`, and forgot to hide this one

File: iota-streams-app/src/transport/tangle/mod.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (arnauorriols) [reply iota-streams-app/src/transport/tangle/mod.rs:232]: I couldn't introduce it without refactoring the wasm-bindings excessively. So I'm still working on it, but I will create a separate PR for it.

[the remark this answers is not in the document]
Comment (arnauorriols) [reply iota-streams-app/src/transport/tangle/mod.rs:232]: PR opened: #140

File: iota-streams-app/src/transport/tangle/mod.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (arnauorriols) [reply iota-streams-app/src/transport/tangle/mod.rs]: done

File: iota-streams-app-channels/src/message/keyload.rs (hunk 17 above)
Comment (arnauorriols) [reply iota-streams-app-channels/src/message/keyload.rs:115]: Ok! My changes are not logical though, conflict resolution should be straightforward :relieved:

File: iota-streams-app-channels/src/api/tangle/user.rs (hunk 14 above)
Comment (arnauorriols) [reply iota-streams-app-channels/src/api/tangle/user.rs:557]: I have been reviewing past commits but they didn't illuminate me. However, my intuition tells me might be something related with the link store?

File: iota-streams-app-channels/src/api/key_store.rs (hunk 12 above)
Comment (arnauorriols) [reply iota-streams-app-channels/src/api/key_store.rs]: Done!

File: bindings/c/src/api/mod.rs (hunk 8 above)
Comment (arnauorriols) [reply bindings/c/src/api/mod.rs:66]: nice suggestion from @semenov-vladyslav

File: iota-streams-app-channels/src/api/user.rs (hunk 16 above)
Comment (arnauorriols) [reply iota-streams-app-channels/src/api/user.rs:1247]: I was wondering where I should place this little thing. Any alternative suggestion?

File: bindings/c/src/api/mod.rs (hunk 9 above)
Comment (arnauorriols) [reply bindings/c/src/api/mod.rs:574]: Yeah... actually the header declarations were already `*const` instead of `*mut`, which makes me wonder... why aren't we using `cbindgen`?

Comment (DyrellC) [reply bindings/c/src/api/mod.rs:574]: I believe we first tried it out with `cbindgen` and ran into some issues, but those issues might have been related to the ternary processes that we've now removed. Haven't tried using it since Oct/Nov of last year.

File: iota-streams-app-channels/src/api/user.rs (hunk 16 above)
Comment (DyrellC) [reply iota-streams-app-channels/src/api/user.rs:1247]: I think here works for now, not sure where else would really be appropriate

File: iota-streams-app-channels/src/api/tangle/user.rs (hunk 14 above)
Comment (DyrellC) [reply iota-streams-app-channels/src/api/tangle/user.rs:557]: Something along those lines I'm guessing. Anyways it's not of any consequence now so we'll use the changed order

File: iota-streams-app/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (DyrellC) [reply iota-streams-app/Cargo.toml:44]: If you wouln't mind updating it to `branch = wasm` for this PR for now, and then we'll update it back once the `iota.rs` wasm bindings pr is merged.

**Review by DyrellC — APPROVED** [review]
Looks good, just switch the branch for `iota.rs` and we can merge.

**Comment by DyrellC** [comment]
There is also the conflict in the `wasm/getting_started.md`

**Comment by arnauorriols** [comment]
> There is also the conflict in the `wasm/getting_started.md`

changed iota-client to wasm branch and rebased onto `develop`.

**Comment by DyrellC** [comment]
Lgtm we should be able to merge now
